lib/zone-spec/fake-async-test.ts

```diff
@@ -188,14 +190,18 @@
 
     flushMicrotasks(): void {
       FakeAsyncTestZoneSpec.assertInZone();
-      while (this._microtasks.length > 0) {
-        let microtask = this._microtasks.shift();
-        microtask();
-        if (this._lastError !== null) {
+      const flushErrors = () => {
+        if (this._lastError !== null || this._uncaughtPromiseErrors.length) {
           // If there is an error stop processing the microtask queue and rethrow the error.
           this._resetLastErrorAndThrow();
         }
       }
+      flushErrors();
+      while (this._microtasks.length > 0) {
+        let microtask = this._microtasks.shift();
+        microtask();
+        flushErrors();
```

Comment:
wouldn't this throw prematurely? I believe the spec-ed behavior and certainly chrome's behavior is that the rejection handler can be registered at any point within the current turn of the event loop:

```
// this does't throw
p = Promise.reject('foo'); Promise.resolve('x').then(() =>  p.then(() => {}, () => {}));


// this throws
p = Promise.reject('foo'); setTimeout(() => { p.then(() => {}, () => {}) }, 0);
```